plugins/inputs/intel_powerstat/services/rapl.go

```diff
@@ -0,0 +1,274 @@
+	dramEnergyPath := fmt.Sprintf(EnergyUjPartialPath, dramPath)
+	dramEnergyFile, err := os.Open(dramEnergyPath)
+	if err != nil {
+		return fmt.Errorf("error opening dram energy_uj file on path %s, err: %v", dramEnergyPath, err)
+	}
+	defer dramEnergyFile.Close()
+
+	return r.calculateData(socketID, energyUjFile, dramEnergyFile)
+}
+
+// GetConstraintMaxPower retrieves ConstraintMaxPower from constraint_0_max_power_uw file.
+func (r *RaplServiceImpl) GetConstraintMaxPower(socketID string) (float64, error) {
+	raplPath := fmt.Sprintf(IntelRaplSocketPartialPath, IntelRaplPath, socketID)
+	maxPowerPath := fmt.Sprintf(MaxPowerUwPartialPath, raplPath)
+	maxPowerFile, err := os.Open(maxPowerPath)
+	if err != nil {
+		return 0, fmt.Errorf("error opening constraint_0_max_power_uw file on path %s, err: %v", maxPowerPath, err)
+	}
+	defer maxPowerFile.Close()
+	maxPower, _, err := r.fs.ReadFileToFloat64(maxPowerFile)
+
+	return maxPower, err
+}
+
+// GetDramMaxEnergyRangeUj retrieves dram energy range value from max_energy_range_uj file.
+func (r *RaplServiceImpl) GetDramMaxEnergyRangeUj(socketID string) (float64, error) {
+	dramPath := fmt.Sprintf(IntelRaplDramPartialPath, IntelRaplPath, socketID, r.dramFolders[socketID])
+	dramEnergyPath := fmt.Sprintf(MaxRangeUjPartialPath, dramPath)
+	maxDramEnergyFile, err := os.Open(dramEnergyPath)
+	if err != nil {
+		return 0, fmt.Errorf("error opening max_energy_range_uj file on path %s, err: %v", dramEnergyPath, err)
+	}
+	defer maxDramEnergyFile.Close()
+	maxDramEnergy, _, err := r.fs.ReadFileToFloat64(maxDramEnergyFile)
+
+	return maxDramEnergy, err
+}
+
+// GetMaxEnergyRangeUj retrieves energy range value from max_energy_range_uj file.
+func (r *RaplServiceImpl) GetMaxEnergyRangeUj(socketID string) (float64, error) {
+	raplPath := fmt.Sprintf(IntelRaplSocketPartialPath, IntelRaplPath, socketID)
+	maxEnergyPath := fmt.Sprintf(MaxRangeUjPartialPath, raplPath)
+	maxEnergyFile, err := os.Open(maxEnergyPath)
+	if err != nil {
+		return 0, fmt.Errorf("error opening max_energy_range_uj file on path %s, err: %v", maxEnergyPath, err)
+	}
+	defer maxEnergyFile.Close()
+	maxEnergy, _, err := r.fs.ReadFileToFloat64(maxEnergyFile)
+
+	return maxEnergy, err
+}
+
+// getDramFolders returns Dram folder names from RAPL.
+func (r *RaplServiceImpl) getDramFolders() map[string]string {
+	return r.dramFolders
+}
+
+func (r *RaplServiceImpl) prepareData() {
+	intelRaplPrefix := "intel-rapl:"
+	intelRapl := fmt.Sprintf("%s%s", intelRaplPrefix, "[0-9]*")
+	raplPaths, err := r.fs.GetStringsMatchingPatternOnPath(fmt.Sprintf("%s/%s", IntelRaplPath, intelRapl))
+	if err != nil {
+		if r.isFirstError {
+			r.isFirstError = false
+			r.log.Errorf("error while preparing RAPL data: %v", err)
+		} else {
+			r.log.Debugf("error while preparing RAPL data: %v", err)
+		}
+
+		r.data = make(map[string]*data.RaplData)
+		return
+	}
+
+	r.isFirstError = true
```

Comment:
I don't see a use of `isFirstError`, you only _write_ this variable but don't read it as far as I can see except for deciding if you want to log an error or a debug message above.

Reply:
Yep, exactly. If you don't have proper directories for reading RAPL data (when it is disabled but plugin will gather other metrics anyway), it will be logged as an error for the first time and on debug level afterwards to make default log nice and clean. If you turn on RAPL when Telegraf is on, it will eventually start gathering metrics from RAPL subsystem.

Reply:
Simplified.